### ISSUE TITLE
Proposed Analyzer: Leap Year Readiness

### ISSUE THREAD

**Issue by mj1856** [dotnet/roslyn-analyzers#2937]
Per step 1 of the [guidelines](https://github.com/dotnet/roslyn-analyzers/blob/master/GuidelinesForNewRules.md), I'm filing an issue in advance of contributing a new analyzer.

My team at Microsoft is focused on Leap Year Readiness, as 2020 is a leap year.  (MSFT employees can visit https://aka.ms/leapyear to learn more.)  As part of this effort, we have built an analyzer for detecting anti-patterns for improper usage of `DateTime` and `DateTimeOffset` that can lead to reliability problems for services running on leap day.  

As an example, our analyzer can detect the following:

```csharp
// problematic code
DateTime result = new DateTime(dt.Year + 1, dt.Month, dt.Day);
```

This matters because when `dt1` is `2020-02-29`, trying to create a date of `2021-02-29` will throw an `ArgumentOutOfRangeException`.

The corrected code should be:

```csharp
// correct code
DateTime result = dt.AddYears(1).Date;
```

Another pattern our analyzer can detect is this:

```csharp
// problematic code
DateTime result = new DateTime(foo.Year, bar.Month, bar.Day);
```

This time, there's no arithmetic going on, but there's still the possibility of `bar` being a Feb 29th, and `foo` being in a year that doesn't have one.  Again, an `ArgumentOutOfRangeException` would be thrown.

There are a few ways to correct for this, but arguably the simplest is:

```csharp
// correct code
DateTime result = bar.AddYears(foo.Year - bar.Year).Date;
```

In all of the above examples, I show `.Date` being called to truncate time, as the original code did not carry through the hours, minutes, seconds, etc.  Also, the same patterns apply to `DateTimeOffset`.

At the moment, we don't have suggested code fixes implemented in the analyzer, and we are still improving our detection logic.  However, we would like to contribute our analyzers to the Roslyn Analyzers repo in their current state while we continue working on improving them.

Since they apply to specific APIs, per the guidelines it seems we shouold be contributing to the `Microsoft.NetCore.Analzyers` project.  I believe we are either in the "Reliability Warnings" or "Usage Warnings" category, but I'd appreciate advice as to which is more appropriate.

Thanks!

**Comment by Evangelink** [comment]
@mj1856 Looks like the leapyear link is not publicly available. Would it be possible to update it? I am genuinely interested about this issue.

**Comment by mj1856** [comment]
@Evangelink - Correct.  That link is to an internal site that goes beyond the scope of this request.

There's not much in the way of official docs in this subject area, other than [the blog post I penned for the 2016 leap year on Azure.com.](https://azure.microsoft.com/en-us/blog/is-your-code-ready-for-the-leap-year/).

I've also done [some personal blogging on this subject recently](https://codeofmatt.com/on-the-life-cycle-of-a-leap-year-bug/), and plan to do more soon.  Thanks for your interest.

**Comment by mavasani** [comment]
Tagging @sharwell and @dotnet/roslyn-analysis for reviewing the proposal from core analysis perspective.

Tagging @stephentoub as the analyzer is about usage of a System.DateTime API.

Overall, the proposal looks fine to me. Please do ensure that if you are not providing code fixes for the initial implementation, then the diagnostic message and documentation link for the analyzer diagnostic has sufficient information for the user to fix the issue. I do think code fixes would be pretty valuable here, especially given the subtle issues you cited in your second problematic code example.

> I believe we are either in the "Reliability Warnings" or "Usage Warnings" category

That sounds good to me - I don't see a problem putting them under either category.

**Comment by stephentoub** [comment]
Sounds fine to me as well at a high-level.

@mj1856, how common are false positives?  Hopefully very rare?

**Comment by mj1856** [comment]
Sorry for the delayed response.

It's difficult to get an exact reading on false positives for an analyzer of this nature.  We do run it across a large amount of internal code, and most of the time it is correctly identifying the anti-pattern.

When we find something that looks like a false positive, we add it as a unit test and work to adjust the analyzer to weed it out.   We probably have a few more of those to add before shipping this, but I'd appreciate the go-ahead to getting a PR started so we can get feedback on our approach and implementation.

More often than not, when we find a false positive, it's not because the pattern didn't match, but because the risk was mitigated by an earlier modification to the expression.  For example:

```csharp
int day = Math.Min(DateTime.DaysInMonth(dt.Year + 1, dt.Month), dt.Day);
DateTime result = new DateTime(dt.Year + 1, dt.Month, dt.Day);
```

Our scanner will hit on the convoluted example above, even though the day has been constrained to the end of the month.   I'm hoping that once we can send what we currently have, that others here can help us figure out how to abate such anomalies.

@kwilkins from our team may be able to provide more specifics.

**Comment by kwilkins** [comment]
Like Matt shared above, the issues we are detecting can be complex when there are modifications prior to the pattern's that we have found to cause issues.

Looking at the example above, although I think `dt.Day` should be `day` within the `DateTime()`
```
int day = Math.Min(DateTime.DaysInMonth(dt.Year + 1, dt.Month), dt.Day);
DateTime result = new DateTime(dt.Year + 1, dt.Month, day);
```

The first line is constraining the day to be leap year safe, which would prevent an issue from happening, but the second line still contains a common pattern that typically does causes an issue. As of now our analyzer hits on the pattern and flags it as a possible leap year causing bug. If this behavior isn't acceptable, we could frame the hit as a suggestion to use the DateTime methods for year increment like so: `DateTime result = dt.AddYears(1);`. This is a good practice aside from leap year, it protects against other bad dates issues too.

The problem we ran into is acceptable lookback detection. In the above example again, we could detect that the `day` variable has been reassigned _in some way_(maybe even keying in on the use of `DateTime.DaysInMonth`) but we haven't been able to always be sure it is leap year safe. If we want to back off more, if we can tell any of the problematic values get altered we could not raise the issue...

But there is also the common pattern when developers incorrectly detect when a leap year is
```
int daysInMonth = month == 2 ? (year % 4 == 0 ? 29 : 28)
  : (month == 4 || month == 6 || month == 9 || month == 11) ? 30
  : 31;
int day = Math.Min(daysInMonth, dt.Day);
DateTime result = new DateTime(dt.Year + 1, dt.Month, day);
```
This example still could result in a leap year issue.

We definitely don't want to surface annoying false positives, but we want to be targeting possible leap year issue causing patterns where we can. Thoughts on this?

**Comment by kwilkins** [comment]
@sharwell and @stephentoub, we have our modifications ready to submit in PR with tests to back it. Do we have approval to move forward?

**Comment by stephentoub** [comment]
It seems reasonable from my perspective.  My concern remains the false positive rate: if there's any liklihood it'll result in a non-trivial number of false positives, I would want the analyzer to be opt-in rather than opt-out.  Otherwise, it'll just end up leading to a lot of noise that will annoy developers into turning it off, and then we're in the worst of all worlds.

**Comment by bgavrilMS** [comment]
Any update on this? Would be great to add this to our open source libraries for continual protection.

**Comment by mj1856** [comment]
@bgavrilMS - The feedback on the PR was that we should rewrite the analyzer using a tainted dataflow analysis approach.  This would be ideal, and would give a much greater accuracy than our current approach.

While we agree, unfortunately we do not have resources to rewrite this to use tainted dataflow analysis in time for it to be released and usable before leap day.

We will still consider working on this to prevent leap year bugs over the long term, as we will have another leap year in 2024, 2028, etc.

Thanks.

**Comment by mj1856** [comment]
Also, I want to point people at an updated blog post on this subject.  Manual review of suspect datetime code is warranted.

https://aka.ms/leapday2020blog

**Comment by carlossanlop** [comment]
Since this issue got approved directly here back in Dec 2019 before we defined that roslyn analyzer proposals should go through API proposal in the runtime repo, and it already has a PR, I'll set it as approved.
